### PR TITLE
Add docker compose support

### PR DESCRIPTION
**Pull request by timcallaghan**
Addresses #172 

Adds support for running Dapr Dashboard inside a docker compose network.

Configuration and Component files are bind mounted to the dashboard container and parsed using similar techniques to the Standalone platform type.

Enumeration of daprd services is done by parsing the docker-compose.yml file and looking for daprd image references.

Metadata fetching is done via first resolving the daprd service location on the docker compose network (using mDNS Dapr component), and then fetching.  Consideration has also been given to the case when the daprd service is protected via an API token.

Readme has been updated with instructions on how to target this new platform type.

### CODE REVIEW

**Comment by timcallaghan** [comment]
Hi @berndverst and @artursouza - I'd love to work through the process of getting this approved and merged.  Do either of you have any capacity to review?  This is my first time using Golang so I'd really appreciate a review and guidance on anything that can be made better 😄

**Comment by Zizico2** [comment]
status?

**Comment by 0GiS0** [comment]
Any news on this? I think is really helpful to see the dashboard in this environment

**Comment by zhupeng0929** [comment]
Any news on this? I think is really helpful to see the dashboard in this environment

**Review by artursouza — APPROVED** [review]

File: pkg/components/components.go [hunk 1]
```diff
@@ -163,3 +168,51 @@ func (c *components) getStandaloneComponents(scope string) []Component {
...
+	dockerComposeComponents := []Component{}
+	err := filepath.Walk(componentsDirectory, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			log.Printf("Failure accessing path %s: %v\n", path, err)
+			return err
+		}
+		if info.IsDir() && info.Name() != filepath.Base(componentsDirectory) {
+			return filepath.SkipDir
+		} else if !info.IsDir() && filepath.Ext(path) == ".yaml" {
+			content, err := ioutil.ReadFile(path)
```
Comment (berndverst) [inline pkg/components/components.go:185]: instead of `ioutil` which is deprecated use the equivalent functions from `os`

**Review by berndverst — CHANGES_REQUESTED** [review]

File: pkg/components/components.go (hunk 1 above)
Comment (artursouza) [reply pkg/components/components.go:185]: I am fixing this now.

File: pkg/components/components.go [hunk 2]
```diff
@@ -14,6 +14,7 @@ limitations under the License.
 package components
 
 import (
+	"io/ioutil"
```
[the remark this answers is not in the document]
Comment (artursouza) [reply pkg/components/components.go:17]: Fixing.

File: cmd/dashboard.go [hunk 3]
```diff
@@ -25,13 +25,39 @@ import (
...
 	dashboardVersion := flag.Bool("version", false, "Prints the dashboard version")
 	port := flag.Int("port", 8080, "Port to listen to")
+	dockerCompose := flag.Bool("docker-compose", false, "Is running inside docker compose")
+	componentsPath := flag.String("components-path", "", "Path to volume mounted dapr components (docker-compose only)")
+	configPath := flag.String("config-path", "", "Path to volume mounted dapr configuration (docker-compose only)")
+	dockerComposePath := flag.String("docker-compose-path", "", "Path to volume mounted docker compose file (docker-compose only)")
 
 	flag.Parse()
 
+	fmt.Println(*dockerCompose)
```
[the remark this answers is not in the document]
Comment (artursouza) [reply cmd/dashboard.go:35]: This is a problem with this project and contributor just followed it. I created this to track this issue: https://github.com/dapr/dashboard/issues/251

**Review by yaron2 — APPROVED** [review]

**Comment by timcallaghan** [comment]
Woo thanks for taking a look at this 🙇

File: cmd/dashboard.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (artursouza) [reply cmd/dashboard.go:35]: We can have another PR if you feel strong about this one.

**Comment by timcallaghan** [comment]
Hi @Mamotchenko - thanks for the very detailed report!

> Have I done something wrong?
Nope, not at all - turns out when I originally created the PR the lib that we're using to load and parse the docker compose file wasn't particularly strict about certain aspects of the compose file. It appears that the lib has been updated and now enforces some stricter constraints - especially one relating to being able to determine the "name" of the compose file, and that in turn requires that the Environment configuration param be set to non-nil.

I recently discovered exactly this bug during the process of writing test coverage, which I've fixed - see https://github.com/dapr/dashboard/pull/252/files#diff-c42e35179b901209bf32a2f7bb15b863e042304762b027b8ea7e81ec8d236bd2

Once the test coverage PR (and bug fix) is approved and merged I think this will solve your issue 🤞 

FYI @artursouza - the code coverage PR needs to be merged to make this actually work with the latest libs we're using.

**Comment by MernaMustafa** [comment]
I've followed the same steps as @Mamotchenko, but I got an empty list for Dapr applications
image

And I've the same error `{"error":"fork/exec /home/nonroot/.dapr/bin/daprd: no such file or directory"}` , It appears in the console but doesn't appear in the dashboard container logs

Any idea why this behavior happens?

**Pull request merged**